### PR TITLE
feat(incentives): implement rollapp gauges

### LINKED ISSUE #945: Implement `rollapp gauges` in the incentives module
add support for `rollapp gauges` to allow streaming incentives to rollapp's sequencer.

PRD: https://www.notion.so/dymension/RollApp-gauge-21e8732a47a24240b473c76d0a16a80e
ADR: https://www.notion.so/dymension/ADR-x-Rollapp-gauges-8b788794ea81427aa44bc5d70112107e

- [x] move incentives module from osmosis repo
 https://github.com/dymensionxyz/dymension/pull/947
- [x] Gauge support for additional `Rollapp Gauge` type (existing type will be called `Asset Gauge`)
- [x] `CreateRollappGauge` keeper method
- [x] distribution to rollapp guage
- [x] rollapp creation hooks
- [x] `RollappGauges` query
- [x] validate if proto migration required
- [ ] dymension hub upgrade handler to create gauges for existing rollapps

### PR DESCRIPTION
**Pull request by mtsitrin**
This PR implements the rollapp gauges feature

- [x] Gauge support for additional Rollapp Gauge type (existing type will be called Asset Gauge)
- [x]  CreateRollappGauge keeper method
- [x]  distribution to rollapp guage
- [x]  rollapp creation hooks
- [x]  RollappGauges query

- [ ] Upgrade handler to create gauges for existing rollapps

Closes #945

### CODE REVIEW

File: x/incentives/keeper/distribute.go [hunk 1]
```diff
@@ -2,41 +2,66 @@ package keeper
...
+			gaugeDistributedCoins sdk.Coins
+			err                   error
+		)
+		switch gauge.DistributeTo.(type) {
+		case *types.Gauge_Asset:
+			gaugeDistributedCoins, err = k.distributeToAssetGauge(ctx, gauge, &lockHolders)
+		case *types.Gauge_Rollapp:
+			gaugeDistributedCoins, err = k.distributeToRollappGauge(ctx, gauge)
+		default:
+			return nil, fmt.Errorf("gauge %d has an unsupported distribution type", gauge.Id)
```
Comment (danwt) [inline x/incentives/keeper/distribute.go:28]: prefer errorsmod.WithType and gerrc.ErrInvalidArgument

File: x/incentives/keeper/gauge_asset.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	"time"
+
+	"github.com/dymensionxyz/dymension/v3/x/incentives/types"
+	lockuptypes "github.com/osmosis-labs/osmosis/v15/x/lockup/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// distributionInfo stores all of the information for pent up sends for rewards distributions.
+// This enables us to lower the number of events and calls to back.
```
Comment (danwt) [inline x/incentives/keeper/gauge_asset.go:14]: calls to back?

File: x/incentives/keeper/gauge_asset.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+// getLocksToDistributionWithMaxDuration returns locks that match the provided lockuptypes QueryCondition,
+// are greater than the provided minDuration, AND have yet to be distributed to.
+func (k Keeper) getLocksToDistributionWithMaxDuration(ctx sdk.Context, distrTo lockuptypes.QueryCondition, minDuration time.Duration) []lockuptypes.PeriodLock {
+	switch distrTo.LockQueryType {
+	case lockuptypes.ByDuration:
+		denom := distrTo.Denom
+		if distrTo.Duration > minDuration {
+			return k.lk.GetLocksLongerThanDurationDenom(ctx, denom, minDuration)
```
Comment (danwt) [inline x/incentives/keeper/gauge_asset.go:42]: correct?

File: x/incentives/keeper/gauge_asset.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+		}
+		return k.lk.GetLocksLongerThanDurationDenom(ctx, distrTo.Denom, distrTo.Duration)
```
Comment (danwt) [inline x/incentives/keeper/gauge_asset.go:44]: can simplify to 1 line with stdlib min or max

File: x/incentives/keeper/gauge_asset.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+	case lockuptypes.ByTime:
+		panic("Gauge by time is present, however is no longer supported. This should have been blocked in ValidateBasic")
+	default:
+	}
+	return []lockuptypes.PeriodLock{}
```
Comment (danwt) [inline x/incentives/keeper/gauge_asset.go:49]: nil

File: x/incentives/keeper/gauge_asset.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+// addLockRewards adds the provided rewards to the lockID mapped to the provided owner address.
+func (d *distributionInfo) addLockRewards(owner string, rewards sdk.Coins) error {
+	if id, ok := d.lockOwnerAddrToID[owner]; ok {
+		oldDistrCoins := d.idToDistrCoins[id]
+		d.idToDistrCoins[id] = rewards.Add(oldDistrCoins...)
+	} else {
+		id := d.nextID
+		d.nextID += 1
```
Comment (danwt) [inline x/incentives/keeper/gauge_asset.go:59]: ++

File: x/incentives/keeper/gauge_asset.go [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+	}
+	return nil
+}
+
+// sendRewardsToLocks utilizes provided distributionInfo to send coins from the module account to various recipients.
+func (k Keeper) sendRewardsToLocks(ctx sdk.Context, distrs *distributionInfo) error {
+	numIDs := len(distrs.idToDecodedAddr)
+	if len(distrs.idToDistrCoins) != numIDs {
+		return fmt.Errorf("number of addresses and coins to distribute to must be equal")
+	}
```
Comment (danwt) [inline x/incentives/keeper/gauge_asset.go:77]: when is this possible?

File: x/incentives/keeper/gauge_asset.go [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+	ctx.Logger().Debug(fmt.Sprintf("Beginning distribution to %d users", numIDs))
```
Comment (danwt) [inline x/incentives/keeper/gauge_asset.go:78]: dont use sprintf in logs, just add the key value pairs

File: x/incentives/keeper/gauge_asset.go [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+	denom := assetDist.Denom
+	lockSum := lockuptypes.SumLocksByDenom(locks, denom)
+
+	if lockSum.IsZero() {
+		return sdk.Coins{}, nil
+	}
+
+	remainCoins := gauge.Coins.Sub(gauge.DistributedCoins...)
+	// if its a perpetual gauge, we set remaining epochs to 1.
+	// otherwise is is a non perpetual gauge and we determine how many epoch payouts are left
```
Comment (danwt) [inline x/incentives/keeper/gauge_asset.go:125]: it's
otherwise it is

File: x/incentives/keeper/gauge_asset.go [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+	remainEpochs := uint64(1)
+	if !gauge.IsPerpetual {
+		remainEpochs = gauge.NumEpochsPaidOver - gauge.FilledEpochs
+	}
```
Comment (danwt) [inline x/incentives/keeper/gauge_asset.go:129]: could be a method on gauge type

File: x/incentives/keeper/gauge_asset.go [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+	/* ---------------------------- defense in depth ---------------------------- */
+	// this should never happen in practice since gauge passed in should always be an active gauge.
+	if remainEpochs == uint64(0) {
```
Comment (danwt) [inline x/incentives/keeper/gauge_asset.go:133]: dont need type conversion

File: x/incentives/keeper/gauge_asset.go [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+		ctx.Logger().Error(fmt.Sprintf("gauge %d has no remaining epochs, skipping", gauge.Id))
+		return sdk.Coins{}, nil
+	}
+
+	// this should never happen in practice
+	if remainCoins.Empty() {
+		ctx.Logger().Error(fmt.Sprintf("gauge %d is empty, skipping", gauge.Id))
+		err := k.updateGaugePostDistribute(ctx, gauge, sdk.Coins{})
+		return sdk.Coins{}, err
```
Comment (danwt) [inline x/incentives/keeper/gauge_asset.go:142]: dont log and return error

File: x/incentives/keeper/grpc_query.go [hunk 13]
```diff
@@ -187,3 +218,29 @@ func (q Querier) filterByPrefixAndDenom(ctx sdk.Context, prefixType []byte, deno
...
+
+// filterRollappGauges
+func (q Querier) filterRollappGauges(ctx sdk.Context, pagination *query.PageRequest) (*query.PageResponse, []types.Gauge, error) {
+	gauges := []types.Gauge{}
+	store := ctx.KVStore(q.Keeper.storeKey)
+	valStore := prefix.NewStore(store, types.KeyPrefixGauges)
+
+	pageRes, err := query.FilteredPaginate(valStore, pagination, func(key []byte, value []byte, accumulate bool) (bool, error) {
+		// this may return multiple gauges at once if two gauges start at the same time.
+		// for now this is treated as an edge case that is not of importance
```
Comment (danwt) [inline x/incentives/keeper/grpc_query.go:230]: not sure I understand second comment

File: x/incentives/types/gauge.go [hunk 14]
```diff
@@ -33,9 +33,18 @@ func NewGauge(id uint64, isPerpetual bool, distrTo lockuptypes.QueryCondition, c
 	}
 }
 
-// IsUpcomingGauge returns true if the gauge's distribution start time is after the provided time.
-func (gauge Gauge) IsUpcomingGauge(curTime time.Time) bool {
-	return curTime.Before(gauge.StartTime)
+// NewGauge creates a new gauge struct given the required gauge parameters.
+func NewRollappGauge(id uint64, rollappId string) Gauge {
```
Comment (danwt) [inline x/incentives/types/gauge.go:37]: wrong docstring

File: x/incentives/types/gauge.go [hunk 15]
```diff
@@ -20,11 +20,11 @@ var (
 )
 
 // NewGauge creates a new gauge struct given the required gauge parameters.
-func NewGauge(id uint64, isPerpetual bool, distrTo lockuptypes.QueryCondition, coins sdk.Coins, startTime time.Time, numEpochsPaidOver uint64, filledEpochs uint64, distrCoins sdk.Coins) Gauge {
+func NewAssetGauge(id uint64, isPerpetual bool, distrTo lockuptypes.QueryCondition, coins sdk.Coins, startTime time.Time, numEpochsPaidOver uint64, filledEpochs uint64, distrCoins sdk.Coins) Gauge {
```
Comment (danwt) [inline x/incentives/types/gauge.go:23]: docstring

**Review by danwt — APPROVED** [review]
I know a lot of it is copy pasted hence the mistakes

File: x/incentives/keeper/gauge_asset.go (hunk 5 above)
Comment (mtsitrin) [reply x/incentives/keeper/gauge_asset.go:48]: it's better this way as the callers doesn't check for != nil

Comment (mtsitrin) [reply x/incentives/keeper/gauge_asset.go:48]: it's better this way as the callers doesn't check for != nil

File: x/incentives/keeper/gauge_asset.go (hunk 12 above)
Comment (mtsitrin) [reply x/incentives/keeper/gauge_asset.go:141]: the `err` is for the `updateGaugePostDistribute`

**Comment by mtsitrin** [comment]
@danwt the logic in `gauge_asset.go` is copy paste from the vanilla module.
I've fixed the low hanging comments.
logic changes, if required, will be done separately

**Pull request merged**